### PR TITLE
8340103

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi!

The old VirtualMemoryTracker has a method set_reserved_region_type(address, flag). We implement this for the new VMATree implementation by altering the signature slightly to set_reserved_region_type(address, size, flag). This simplifies the implementation greatly for our new data structure and leads to trivial changes for the callers (all callers already know the size).

This PR implements the internal implementation along with tests, but does not change any callers.

I also do a few cleanups:

- Change `Node` to `TreeNode` in tests, we've seen build failures because of this (probably a precompiled headers issue)
- Add a few `print_on` methods for  easy debugging
- Add a `size` alias, it was a bit confusing that some functions took an argument `position sz`, so changed that to `size sz` 

Thanks.

### CODE REVIEW

File: src/hotspot/share/nmt/vmatree.cpp [hunk 1]
```diff
@@ -197,3 +197,57 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
...
+               statetype_to_string(current->val().out.type()));
+  });
+  out->cr();
+}
+#endif
+
+VMATree::SummaryDiff VMATree::set_tag(position from, size size, MemTag tag) {
+  VMATreap::Range range = _tree.find_enclosing_range(from);
+  assert(range.start != nullptr && range.end != nullptr,
+         "Setting a flag must be done within existing range");
```
Comment (afshin-zafari) [inline src/hotspot/share/nmt/vmatree.cpp:214]: flag --> tag

File: src/hotspot/share/nmt/vmatree.cpp [hunk 2]
```diff
@@ -197,3 +197,57 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
...
+  StateType type = range.start->val().out.type();
+  RegionData new_data = RegionData(range.start->val().out.stack(), tag);
+
+  position end = MIN2(from + size, range.end->key());
+  SummaryDiff diff = register_mapping(from, end, type, new_data);
+  size = size - (end - from);
+  from = end;
+
+  // If end < from + sz then there are multiple ranges for which to set the flag.
```
Comment (afshin-zafari) [inline src/hotspot/share/nmt/vmatree.cpp:223]: flag --> tag

File: src/hotspot/share/nmt/vmatree.cpp [hunk 3]
```diff
@@ -197,3 +197,57 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
...
+  while (end < from + size) {
+    range = _tree.find_enclosing_range(from);
+    assert(range.start != nullptr && range.end != nullptr,
+           "Setting a flag must be done within existing range.");
```
Comment (afshin-zafari) [inline src/hotspot/share/nmt/vmatree.cpp:227]: flag --> tag

**Review by afshin-zafari — COMMENTED** [review]
Thanks for the work. This enables the `MemTracker::record_virtual_memory_tag` to just use this API and apply the results.

Just a few cases of 'flag' is missed in comments or in assertion strings. There should be 'tag'.
Can we use `add(SummaryDiff& other)` instead of `apply(SummaryDiff& other)`?

**Comment by jdksjolen** [comment]
> Thanks for the work. This enables the `MemTracker::record_virtual_memory_tag` to just use this API and apply the results.
> 
> Just a few cases of 'flag' is missed in comments or in assertion strings. There should be 'tag'. Can we use `add(SummaryDiff& other)` instead of `apply(SummaryDiff& other)`?

Done.

**Review by afshin-zafari — APPROVED** [review]
Thanks.

File: src/hotspot/share/nmt/vmatree.cpp [hunk 4]
```diff
@@ -197,3 +197,57 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
   }
   return diff;
 }
+
+#ifdef ASSERT
+void VMATree::print_on(outputStream* out) {
+  visit_in_order([&](TreapNode* current) {
+    out->print(SIZE_FORMAT " (%s) - %s - ", current->key(), NMTUtil::tag_to_name(current->val().out.mem_tag()),
+               statetype_to_string(current->val().out.type()));
```
Comment (gerard-ziemski) [inline src/hotspot/share/nmt/vmatree.cpp:205]: Could we consider here adding a utility function:

```
  IntervalState getState(TreapNode* node) {
    return node->val().out;
  }
```

to simplify from:

```
    out->print(SIZE_FORMAT " (%s) - %s - ", current->key(), NMTUtil::tag_to_name(current->val().out.mem_tag()),
                statetype_to_string(current->val().out.type()));
```

to

```
    out->print(SIZE_FORMAT " (%s) - %s - ", current->key(), NMTUtil::tag_to_name(getState(current).mem_tag()),
                statetype_to_string(getState(current).type()));
```

and all the other from:

`    stB.out = leqA_n->val().out;
`

to:

`    stB.out = getState(leqA_n);
`

Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.cpp]: Sure, but I had to pick `in_state` and `out_state` as there are two states we're interested in.

**Review by afshin-zafari — APPROVED** [review]
Still looks good to me. Thanks

**Review by gerard-ziemski — APPROVED** [review]
LGTM

**Comment by jdksjolen** [comment]
@afshin-zafari , @gerard-ziemski 

Please re-review this PR. I've added an early-return if the range isn't found.

File: src/hotspot/share/nmt/vmatree.cpp [hunk 5]
```diff
@@ -215,3 +216,63 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
...
+  if (range.start == nullptr || range.end == nullptr) {
+    log_debug(nmt)("set_tag called on non-existent range: " PTR_FORMAT " - " PTR_FORMAT,
+                   from, from+size);
+    return SummaryDiff();
+  }
+
+  StateType type = out_state(range.start).type();
+  RegionData new_data = RegionData(out_state(range.start).stack(), tag);
+
+  position end = MIN2(from + size, range.end->key());
```
Comment (tstuefe) [inline src/hotspot/share/nmt/vmatree.cpp:243]: I get that `key` is the address which is the key for the treap, but using "key" at places where we use the address is super confusing for the casual code reader. Can we give TreapNode an alias, for "key"? E.g. "address" or "position"?

File: src/hotspot/share/nmt/vmatree.cpp [hunk 6]
```diff
@@ -215,3 +216,63 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
...
+VMATree::SummaryDiff VMATree::set_tag(position from, size size, MemTag tag) {
+  VMATreap::Range range = _tree.find_enclosing_range(from);
+  assert(range.start != nullptr && range.end != nullptr,
+         "Setting a memory tag must be done within existing range");
+  if (range.start == nullptr || range.end == nullptr) {
+    log_debug(nmt)("set_tag called on non-existent range: " PTR_FORMAT " - " PTR_FORMAT,
+                   from, from+size);
+    return SummaryDiff();
+  }
+
```
Comment (tstuefe) [inline src/hotspot/share/nmt/vmatree.cpp:239]: Please add a short two-line comment here describing the algorithm.

File: src/hotspot/share/nmt/vmatree.cpp [hunk 7]
```diff
@@ -215,3 +216,63 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
...
+  StateType type = out_state(range.start).type();
+  RegionData new_data = RegionData(out_state(range.start).stack(), tag);
+
+  position end = MIN2(from + size, range.end->key());
+  SummaryDiff diff = register_mapping(from, end, type, new_data);
+  size = size - (end - from);
```
Comment (tstuefe) [inline src/hotspot/share/nmt/vmatree.cpp:245]: Can you please not modify the input parameters, instead use temporary local vars? E.g. "pos" and "remaining".

File: src/hotspot/share/nmt/vmatree.cpp [hunk 8]
```diff
@@ -215,3 +216,63 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
...
+  from = end;
+
+  // If end < from + sz then there are multiple ranges for which to set the flag.
+  while (end < from + size) {
+    range = _tree.find_enclosing_range(from);
```
Comment (tstuefe) [inline src/hotspot/share/nmt/vmatree.cpp:250]: This is a bit of a brain teaser.

I first wondered why we needed to find the "from" node again, since we already have it. But the last `register_mapping` call can actually have coalesced ranges and either the old `from` node or the old `to` node can have disappeared. 

So, we have two possibilities: 
- the old "to" node is still there, in which case we could just read its RegionData state and use that for the next segment (modulo tag)
- the old "to" node disappeared due to merging (meaning, tagging the last segment causes it to coalesce with the next segment), in which case there is nothing to do.

But I am not 100% sure. So, re-querying the range from the treap is probably the safest way to go. 

But can you please add a comment stating that register_mapping may have modified the tree structure and caused segments to coalesce, just invalidating the old `range` ?

File: test/hotspot/gtest/nmt/test_vmatree.cpp [hunk 9]
```diff
@@ -284,6 +284,68 @@ TEST_VM_F(NMTVMATreeTest, LowLevel) {
...
+TEST_VM_F(NMTVMATreeTest, SetTag) {
+  auto i = [](MemTag f) -> uint8_t { return (uint8_t)f; };
+
+  // The gc/cds case with only reserved data
+  {
+    VMATree::SummaryDiff diff;
+    Tree::RegionData rd(NCS::StackIndex(), mtNone);
+    VMATree tree;
+
+    VMATree::SummaryDiff result = tree.reserve_mapping(0, 500, rd);
```
Comment (tstuefe) [inline test/hotspot/gtest/nmt/test_vmatree.cpp:296]: How does using '0' gel with your find_enclosing_range, where NULL for start or end is "did not find"?

File: test/hotspot/gtest/nmt/test_vmatree.cpp [hunk 10]
```diff
@@ -284,6 +284,68 @@ TEST_VM_F(NMTVMATreeTest, LowLevel) {
...
+    result = tree.set_tag(0, 500, mtGC);
+    diff.add(result);
+    EXPECT_EQ(100, diff.tag[i(mtNone)].reserve);
+    EXPECT_EQ(500, diff.tag[i(mtGC)].reserve);
+
+    result = tree.set_tag(500, 100, mtClassShared);
+    diff.add(result);
+    EXPECT_EQ(0, diff.tag[i(mtNone)].reserve);
+    EXPECT_EQ(500, diff.tag[i(mtGC)].reserve);
+    EXPECT_EQ(100, diff.tag[i(mtClassShared)].reserve);
```
Comment (tstuefe) [inline test/hotspot/gtest/nmt/test_vmatree.cpp:313]: As  it is now, the test only compares sums, and needs quite a bit of boilerplate coding for that.

My proposal would be:

- give VMATree a test function that takes an array of ranges (plain simple stupid POD structs) and compares them item by item with its internal state. Let it return true on success, print an error on an outputStream* we give the function on error.
- then, in the test, hardcode the test array and let VMATree compare it.

Example call for above test:
```
struct testrange { from, to, stackindex, tag };
const testrange[] expected_form = {
 { 0, 500, NCS::StackIndex(), mtGC },
 { 500, 600, NCS::StackIndex(), mtClassShared },
};
tree.test_against(expected_form, tty);
```

That would be easier on the eye in the test file, more succinct, and catch more errors. Because we compare actual addresses, and we also compare expected stack indices. As it is now, this test would not catch if we accidentally screw up stacks.

File: test/hotspot/gtest/nmt/test_vmatree.cpp [hunk 11]
```diff
@@ -284,6 +284,68 @@ TEST_VM_F(NMTVMATreeTest, LowLevel) {
...
+    EXPECT_EQ(500, diff.tag[i(mtGC)].reserve);
+    EXPECT_EQ(125, diff.tag[i(mtGC)].commit);
+
+    result = tree.set_tag(500, 100, mtClassShared);
+    diff.add(result);
+    EXPECT_EQ(0, diff.tag[i(mtNone)].reserve);
+    EXPECT_EQ(100, diff.tag[i(mtClassShared)].reserve);
+    EXPECT_EQ(20, diff.tag[i(mtClassShared)].commit);
+  }
+}
```
Comment (tstuefe) [inline test/hotspot/gtest/nmt/test_vmatree.cpp:347]: I would like to see the following cases explicitly tested:
1) set a tag over a range spanning multiple existing segments with the same stack but different tags - here, I would like to see that set_tag causes the tree to coalesce these different ranges to a single range
2) set a tag over a range spanning multiple existing segments with different stacks and different tags - here, I would like to see that the original ranges stay intact, including their stack information, only the tags should now all be the same
3) set a tag within a range - here, I would like to see three ranges now

for 1, also for mismatched starts and ends where either start or end are in the middle of an existing range, which should cause the tree to split that range

**Review by tstuefe — COMMENTED** [review]
I slowly am getting up for air after Lilliput. Sorry for the dragged-out reviews.

File: src/hotspot/share/nmt/nmtTreap.hpp [hunk 12]
```diff
@@ -304,6 +304,36 @@ class Treap {
...
+    TreapNode* end;
+  };
+
+  // Return the range [start, end)
+  // where start->key() <= addr < end->key().
+  // Failure to find the range leads to start and/or end being null.
+  Range find_enclosing_range(K addr) {
+    TreapNode* start = closest_leq(addr);
+    TreapNode* end = closest_gt(addr);
+    return Range{start, end};
```
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/nmt/nmtTreap.hpp:334]: I'd just do curly brace initialization the same way as paren init. I don't think our style guide says anything, but the typical style that I see in C++ code is this dense variant.

File: src/hotspot/share/nmt/vmatree.hpp [hunk 13]
```diff
@@ -140,6 +142,14 @@ class VMATree {
 private:
   VMATreap _tree;
 
+  IntervalState& in_state(TreapNode* node) {
+    return node->val().in;
+  }
+
+  IntervalState& out_state(TreapNode* node) {
+    return node->val().out;
+  }
```
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.hpp:151]: They should be static and const, yes.

Re: ephemeral utility class. I do prefer using functions as opposed to methods, but it'd be nice to wrap them in a namespace or class that can be shortened and used outside of `VMATree`.

File: src/hotspot/share/nmt/nmtTreap.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (tstuefe) [reply src/hotspot/share/nmt/nmtTreap.hpp:334]: But it is not uniform if we end up with three different initialization styles: the old way of `object(a, b, c);` the new way of `object{a, b, c};` and plain array/struct init, which  overwhelmingly seems to use spaces around curlies: `= { a, b, c };` .

File: src/hotspot/share/nmt/vmatree.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (tstuefe) [reply src/hotspot/share/nmt/vmatree.hpp:151]: Okay

File: src/hotspot/share/nmt/vmatree.cpp [hunk 14]
```diff
@@ -215,3 +216,63 @@ VMATree::SummaryDiff VMATree::register_mapping(position A, position B, StateType
...
+#ifdef ASSERT
+void VMATree::print_on(outputStream* out) {
+  visit_in_order([&](TreapNode* current) {
+    out->print(SIZE_FORMAT " (%s) - %s - ", current->key(), NMTUtil::tag_to_name(out_state(current).mem_tag()),
+               statetype_to_string(out_state(current).type()));
+  });
+  out->cr();
+}
+#endif
+
```
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.cpp:229]: >what if there are address holes in that range?

That's a good point. The function mainly exists to provide support for `record_virtual_memory_tag`, so I haven't thought too hard about the edge cases. In the case of `Released` nodes, I think we're still safe in that we don't look at the remaining state of the node. Anyway, some docs is in order.

File: test/hotspot/gtest/nmt/test_vmatree.cpp (hunk 9 above)
Comment (jdksjolen) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:296]: Sorry, I don't see what the relevance is for this line of code with the rest of the PR :-). Of course, `nullptr` is a pointer type so can't be passed in as a numerical argument, as opposed to `NULL`.

File: src/hotspot/share/nmt/vmatree.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (tstuefe) [reply src/hotspot/share/nmt/vmatree.cpp:229]: An interesting point to think about (I did not look into the code): released sections should always coalesce. So they always should have the same properties (null stack and mtNone). I'm pretty sure we do this already.

But we must then make sure we don't set the tag for released sections.

We also could just say set_tag is not allowed in released sections, and if the range spans over an address hole, just assert. In the way VMATree is used, that is almost certainly an error.

[the remark this answers is not in the document]
Comment (tstuefe) [reply src/hotspot/share/nmt/vmatree.cpp:229]: BTW in case that is not clear, I love that you do this PR. This is exactly what I had hoped for (retagging existing mapped ranges).

**Comment by jdksjolen** [comment]
Hi Thomas,

Thanks for the good feedback! It was a while since I worked on this PR, so the round trip time for review+response is going to be a bit longer than usual.

Cheers

File: src/hotspot/share/nmt/vmatree.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.hpp:151]: >Also, TreapNode* should be const.

Right, gotta back off from that again. The issue is that `val()` is non-const, which is correct: we should be able to change the value of a node. But, that means that this argument cannot be `const`, because (quoting my compiler`: `'this' argument discards qualifiers`. It's just an unfortunately clunky part of C++, it seems.

File: test/hotspot/gtest/nmt/test_vmatree.cpp (hunk 9 above)
Comment (tstuefe) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:296]: Ah, my mistake. Range.end/start is null, so the TreapNode* itself is null, not the address the TreapNode* represents.

File: test/hotspot/gtest/nmt/test_vmatree.cpp (hunk 11 above)
Comment (jdksjolen) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:418]: Tests added, and they all passed! I really didn't think that case 3 would work, but sometimes I write good code I guess.

File: src/hotspot/share/nmt/vmatree.cpp (hunk 8 above)
Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.cpp:252]: Comment added.

File: src/hotspot/share/nmt/nmtTreap.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/nmt/nmtTreap.hpp:338]: Yeah, I don't care too much about what we do here to be honest. If you want spaces around curlies, we can do that :).

File: test/hotspot/gtest/nmt/test_vmatree.cpp [hunk 15]
```diff
@@ -284,6 +285,138 @@ TEST_VM_F(NMTVMATreeTest, LowLevel) {
...
+    VMATree tree;
+    Tree::RegionData gc(si, mtGC);
+    Tree::RegionData compiler(si, mtCompiler);
+    tree.reserve_mapping(0, 100, gc);
+    tree.reserve_mapping(100, 100, compiler);
+    tree.set_tag(75, 50, mtClass);
+    expect_equivalent_form(expected, tree);
+  }
+}
+
```
Comment (tstuefe) [inline test/hotspot/gtest/nmt/test_vmatree.cpp:419]: Do we allow address holes? If yes, please add a test where we set_tag over an address hole. If the expected behavior is to ignore the hole, this should be easily tested.

Otherwise, if we assert in case of set_tag over a hole, its fine. We could write a death test to test the assert, but I don't think that is worth the test cycles (death test are expensive).

**Review by tstuefe — COMMENTED** [review]
Tests: very nice. Comments inline.

File: test/hotspot/gtest/nmt/test_vmatree.cpp (hunk 11 above)
Comment (tstuefe) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:418]: That's the good thing about tests. Gives you confidence.

File: test/hotspot/gtest/nmt/test_vmatree.cpp [hunk 16]
```diff
@@ -415,6 +422,19 @@ TEST_VM_F(NMTVMATreeTest, SetTag) {
...
+        { 0,  50,          mtGC, si, State::Reserved},
+        {50,  75,          mtGC, si, State::Released},
+        {75, 100, mtClassShared, si, State::Reserved}
+    };
+    VMATree tree;
+    Tree::RegionData class_shared(si, mtClassShared);
+    tree.reserve_mapping(0, 100, class_shared);
+    tree.set_tag(0, 75, mtGC);
+    expect_equivalent_form(expected, tree);
+  }
```
Comment (tstuefe) [inline test/hotspot/gtest/nmt/test_vmatree.cpp:437]: I think this shows an error I mentioned before, which should be fixed:

Released regions should only have mtNone as tag and no associated stack. The reason is that we don't want released sections to splinter into different parts, since that makes no sense. Released sections should always coalesce and that is easiest reached by making its properties uniformly "mtNone, no stack"

Comment (jdksjolen) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:439]: You might be right about that. I thought that we simply ignored state and stack of released regions. Let me see what's going on.

Comment (jdksjolen) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:439]: Ah, actually that is precisely what we see as an assert later on. `set_tag` must support address holes in some sense, otherwise we'll end up in a bad state.

File: src/hotspot/share/nmt/vmatree.cpp (hunk 7 above)
Comment (jdksjolen) [reply src/hotspot/share/nmt/vmatree.cpp:251]: Sure

**Comment by jdksjolen** [comment]
@tstuefe , alright I've cleaned up a bit and fixed the tests. For address holes, the solution is to skip released ranges. I think I want another test for that case, so I'll make sure to get that up. Other than that, the code is pretty good. I guess I could write a small explanation of the algorithm used in `set_tag`, if you find it unintuitive?

File: test/hotspot/gtest/nmt/test_vmatree.cpp (hunk 16 above)
Comment (tstuefe) [reply test/hotspot/gtest/nmt/test_vmatree.cpp:439]: I think register_mapping should always register released sections with mtNone/no stack. Maybe even assert if you specify something different. Then, set_tag should just ignore released sections, not change anything

**Comment by jdksjolen** [comment]
Mac OSX failure unrelated to PR.

**Comment by tstuefe** [comment]
> @tstuefe , alright I've cleaned up a bit and fixed the tests. For address holes, the solution is to skip released ranges. I think I want another test for that case, so I'll make sure to get that up. Other than that, the code is pretty good. I guess I could write a small explanation of the algorithm used in `set_tag`, if you find it unintuitive?

Not an explanation of the algorithm (although you can if you like) but an explanation of the expected behavior above the method in the header.

"Given an NMT tag and a memory range, changes the tag associated with that memory range. The memory range does not have to fall into existing memory boundaries (setting tags for just parts of existing sections is valid). The memory range may overlap multiple existing sections, but sections in Released state are ignored (Released sections are always tagged with mtNone). This operation may cause existing sections to coalesce or new sections to form".

Then maybe two or three examples of coalescing and splitting.

**Comment by tstuefe** [comment]
@jdksjolen 
" Last test didn't actually make any sense Since the VMATree will remove the released range when it recognises that it is a no-op"

not so hasty :) The test was fine. You said that set_tag just ignores Released sections in range. So, what should have happened is that in the test, the remaining reserved section should have remained and changed to mtGC. The tree then should consist of `[0, 50) Released, mtNone, nostack) [50, 75) mtGC, si) [75, 100) Released, mtNone, nostack)`. 

In other words, the test would test that calling set_tag across a range that contains Released sections does not change said sections.

**Comment by jdksjolen** [comment]
> @jdksjolen " Last test didn't actually make any sense Since the VMATree will remove the released range when it recognises that it is a no-op"
> 
> not so hasty :) The test was fine. You said that set_tag just ignores Released sections in range. So, what should have happened is that in the test, the remaining reserved section should have remained and changed to mtGC. The tree then should consist of `[0, 50) Released, mtNone, nostack) [50, 75) mtGC, si) [75, 100) Released, mtNone, nostack)`.
> 
> In other words, the test would test that calling set_tag across a range that contains Released sections does not change said sections.

I actually am stuck thinking about this, and it's worse than that :-).

Consider a simpler test:

```c++
{
    VMATree tree;
    Tree::RegionData class_shared(si, mtClassShared);
    tree.reserve_mapping(10, 10, class_shared);
    tree.set_tag(0, 100, mtCompiler);
}
```

This will crash on an assert. After the `reserve_mapping` the state of the tree is `Res,10 - Res,20`. That is: Two reserved nodes, at position 10 and 20.

So, when we run `find_enclosing_range(0)`, we get a non-existent range back, because no range encloses `0`. I'm pretty sure that we can have better behavior by interpreting the piece-wise results of `find_enclosing_range`.

**Comment by jdksjolen** [comment]
> > > @jdksjolen " Last test didn't actually make any sense Since the VMATree will remove the released range when it recognises that it is a no-op"
> > > not so hasty :) The test was fine. You said that set_tag just ignores Released sections in range. So, what should have happened is that in the test, the remaining reserved section should have remained and changed to mtGC. The tree then should consist of `[0, 50) Released, mtNone, nostack) [50, 75) mtGC, si) [75, 100) Released, mtNone, nostack)`.
> > > In other words, the test would test that calling set_tag across a range that contains Released sections does not change said sections.
> > 
> > 
> > I actually am stuck thinking about this, and it's worse than that :-).
> > Consider a simpler test:
> > ```c++
> > {
> >     VMATree tree;
> >     Tree::RegionData class_shared(si, mtClassShared);
> >     tree.reserve_mapping(10, 10, class_shared);
> >     tree.set_tag(0, 100, mtCompiler);
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This will crash on an assert. After the `reserve_mapping` the state of the tree is `Res,10 - Res,20`. That is: Two reserved nodes, at position 10 and 20.
> > So, when we run `find_enclosing_range(0)`, we get a non-existent range back, because no range encloses `0`. I'm pretty sure that we can have better behavior by interpreting the piece-wise results of `find_enclosing_range`.
> 
> I wonder:
> 
> The answer could be to start the tree out with a single range, State=Released, mtNone, nostack, from [0 .. max). In other words, have two nodes that border the very start and end of the `position` value range.
> 
> The only problematic part is that one must encode the logic that there cannot be a node preceding the first node - so it has no input state (or it can have one, but should be ignored).
> 
> What do you think? Elegant or too overengineered?

You could alter the `VMATree` to look like that. I'd rather have fewer assumptions made by `VMATree`, and a more complicated `set_tag`. The function can deal with these challenges, no problem.

**Comment by tstuefe** [comment]
Hey @jdksjolen, hier is a thougt. Maybe for this RFE, maybe for future rework:

We now add a function set_tag to modify the tag across a range of regions. That requires us to preserve the rest of the properties. So, you essentially go through the sections, read old properties, modify this one property, then overwrite the old section with the new section with the changed properties. The tree does the rest (coalescing, splitting etc).

But it occurred to me, that we may need this functionality for other properties, too. For example, State: We may want to implement "uncommit" as "uncommit across possibly multiple sections that may or may not have been committed, but preserve tags and stacks".

So, we have a general need for "change properties across a range of positions" but where we don't just wipe out the old properties but modify them. So we need to know the old set.

How about we implement this directly in the treap in a more general purpose way. For example:

`void transform(Transformer x, positon from, position to)`

with Transformer being a functor that gets the old property set (and possibly the positions of the old range, though that may not even be needed) and returns a new set of properties.

The VMATree transformer for VMATree::set_tag would just return the input properties, but with Tag replaced by the new tag.

A VMATree transformer for a possible "commit" function would return the input properties, but with State=Comitted. So, we could commit over a range of existing sections with different tags or stacks, and that would preserve the stacks and tags.

And so forth. What do you think?



As in the case for set_tag: the operation is "Sort of "transform" operation.

Until now, the treap had the simple operation of "register region" where we register a region with a

**Comment by jdksjolen** [comment]
> Hey @jdksjolen, hier is a thougt. Maybe for this RFE, maybe for future rework:
> 
> We now add a function set_tag to modify the tag across a range of regions. That requires us to preserve the rest of the properties. So, you essentially go through the sections, read old property set, modify this one property, then exchange the old property set for this section with the new property set. The tree does the rest automagically (coalescing, splitting etc).
> 
> But it occurred to me, that we may need this functionality for other properties, too. For example, State: We may want to implement "uncommit" as "uncommit across possibly multiple sections that may or may not have been committed, but preserve tags and stacks".
> 
> So, we have a general need for "change properties across a range of positions" but where we don't just wipe out the old properties but modify them. So we need to know the old set.
> 
> How about we implement this directly in the treap in a more general purpose way. For example:
> 
> `void transform(Transformer x, positon from, position to)`
> 
> with Transformer being a functor that gets the old property set (and possibly the positions of the old range, though that may not even be needed) and returns a new set of properties. `transform` would call the `Transformer` for every section it encounters.
> 
> The VMATree transformer for VMATree::set_tag would just return the input properties, but with Tag replaced by the new tag.
> 
> A VMATree transformer for a possible "commit" function would return the input properties, but with State=Comitted. So, we could commit over a range of existing sections with different tags or stacks, and that would preserve the stacks and tags.
> 
> And so forth. What do you think?

I think that this is probably a generalization of the algorithm in `register_mapping`, or quite similar to it. Sure, that can be very useful, if it's needed. I do think that that is for a future RFE, however.

**Comment by jdksjolen** [comment]
Hi,

I've updated the PR with further tests and a more complete implementation of `set_flag`. In a follow up RFE, we can probably extract this to some sort of `transfomer` function.

**Comment by jdksjolen** [comment]
Thank you!
/integrate
